### ISSUE TITLE
nathelper: set_contact_alias is not applied

### ISSUE THREAD

**Issue by oej** [kamailio/kamailio#445]
nathelper module: If I run set_contact_alias nothing happens - the contact remains the same on the outbound message. There's no error returned and debug logs indicate a proper header. 

However, if I add "msg_apply_changes()" on the line after set_contact_alias the Contact sent to next hop has the alias content. No other changes.

**Comment by oej** [comment]
I see the same issue with add_contact_alias()

**Comment by miconda** [comment]
IIRC, add_contact_alias() doesn't add alias if the address is contact is the same as the address to be added as alias.

You description about the behavior of set_contact_alias() is not clear for me. Maybe you can paste here the incoming and outgoing requests when using set contact alias.

**Comment by oej** [comment]
I send in a contact that is a 10.10.x.x address - clearly not the same as $si that is in 192.168.40.x (also tested on public IP for server) - and both functions return true, indicating no failure. The outbound message keeps the same contact as the inbound message.

If I run msg_apply_changes() the alias is added to the contact for the outbound message as it should be.

Very simple configuration, receiving message, running set_contact_alias() and then issuing t_relay().

**Comment by miconda** [comment]
I couldn't track any change on the code used by set_contact_alias(). Are you using the function in main request_route or a branch_route/etc.?

**Comment by oej** [comment]
Main request route.

**Comment by miconda** [comment]
If you have a simple config, can you paste it here, at least the request_route and any branch_route used for this case (also onsend_route if in config).

**Comment by oej** [comment]
´´´
# -------------------------  request routing logic -------------------
request_route {
   xlog("L_INFO", "--> Incoming request: $rm from $si to $ru \n$mb\n");

   # Always try to fix contact alias if needed
   if(!set_contact_alias()) {
	sl_send_reply("400", "Contact alias failed");
	xlog("L_ERR", "---- Contact alias failed: $ct\n");
   }
   #msg_apply_changes();

   if(!t_relay()) {
      # Transaction failed, send stateless reply
      xlog("L_INFO", "   --> Transaction failed.  Sending error response. \n");
      sl_reply_error();
   }
   exit;
}

onsend_route
{
        xlog("L_INFO", " --->ONSEND from $sndfrom(ip) to $snd(ip) \n$mb\n---\n");
}
´´´

**Comment by miconda** [comment]
I tested with master branch on a config based on the default one and all is ok. Maybe you can add some debug messages in the source code and see how far the code in the alias function is executed.

**Comment by miconda** [comment]
Tested also with your config snippet and all look ok. I was sending a SIP MESSAGE request, being the one I could use for testing in my current environment, but the code is the same.

**Comment by oej** [comment]
I had a onsend-route in my original setup, maybe that affected the outcome, but I can't see how. I only get the parameters when using msg_apply_changes in both 4.3 and trunk.

**Comment by miconda** [comment]
I used onsend route also in the second test, which is your snipped above, plus loading the required modules.

**Comment by miconda** [comment]
Any update on this one, not being reproducible here. As suggested above, maybe adding some log messages in the source code of the function will help getting more hints.

**Comment by oej** [comment]
If we get time at fosdem, I'll show you :-)

**Comment by miconda** [comment]
I would like to sort this item in a way or another. It doesn't show in tests.

**Comment by oej** [comment]
I will reopen when I have time and resources to get back on this.

**Comment by jbj73** [comment]
Hello, 
I ran in to the same issue where set_contact_alias were applied in some cases and in some cases not. I nailed it down to be related to having GRUU support enabled in the registrar.

Originally I had the configuration:
modparam("registrar", "gruu_enabled", 1)

Disabling GRUU made it work as I expected:
modparam("registrar", "gruu_enabled", 0)

This may be expected behaviour but in my case I tried to figure out why it in some cases did apply the alias and in other cases not. Maybe this will save someone else a few hours of confusion and frustration...
